projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py

```diff
@@ -0,0 +1,109 @@
+            ),
+            randomize_delay_seconds=int(
+                os.environ.get(
+                    "VDK_META_JOBS_DELAYED_JOBS_RANDOMIZED_ADDED_DELAY_SECONDS", 600
+                )
+            ),
+        )
+        self._finished_jobs = []
+        self._job_executor = TrackingDataJobExecutor(RemoteDataJobExecutor())
+
+    def build_dag(self, jobs: List[Dict]):
+        for job in jobs:
+            trackable_job = TrackableJob(
+                job["job_name"],
+                job.get("team_name", self._team_name),
+                job.get("fail_meta_job_on_error", True),
+            )
+            self._job_executor.register_job(trackable_job)
+            self._topological_sorter.add(trackable_job.job_name, *job["depends_on"])
+
+    def execute_dag(self):
+        self._topological_sorter.prepare()
+        while self._topological_sorter.is_active():
+            for node in self._topological_sorter.get_ready():
+                self._start_job(node)
+            self._start_delayed_jobs()
+
+            for node in self._get_finalized_jobs():
+                if node not in self._finished_jobs:
+                    log.info(f"Data Job {node} has finished.")
+                    self._topological_sorter.done(node)
+                    self._job_executor.finalize_job(node)
+                    self._finished_jobs.append(node)
+
+    def _start_delayed_jobs(self):
+        while True:
+            job = self._delayed_starting_jobs.dequeue()
+            if job is None:
+                break
+            log.info(f"Trying to start job {job} again.")
+            self._start_job(job)
+
+    def __repr__(self):
+        # TODO move out of this class
+        def default_serialization(o: Any) -> Any:
+            return o.__dict__ if "__dict__" in dir(o) else str(o)
+
+        jobs = self._job_executor.get_all_jobs()
+        data = [j.details if j.details is not None else j.__dict__ for j in jobs]
+        try:
+            result = json.dumps(data, default=default_serialization, indent=2)
+        except Exception as e:
+            log.debug(f"Failed to json.dumps : {e}. Fallback to pprint.")
+            # sort_dicts is supported since 3.8
+            if sys.version_info[0] >= 3 and sys.version_info[1] >= 8:
+                result = pprint.pformat(
+                    data, indent=2, depth=5, compact=False, sort_dicts=False
+                )
+            else:
+                result = pprint.pformat(data, indent=2, depth=5, compact=False)
+        return result
+
+    def _start_job(self, node):
+        try:
+            self._job_executor.start_job(node)
+        except ApiException as e:
+            if e.status == 409:
+                log.info(
+                    f"Detected conflict with another running job: {e}. Will be re-tried later"
+                )
+                self._delayed_starting_jobs.enqueue(node)
+            elif e.status >= 500:
+                log.info(
```

Comment:
should this be an error?

Reply:
No, the job will be re-tried. We do not log errors on re-tries.